src/mds/mdstypes.h

```diff
@@ -1170,6 +1170,7 @@ struct session_info_t {
   entity_inst_t inst;
   std::map<ceph_tid_t,inodeno_t> completed_requests;
   interval_set<inodeno_t> prealloc_inos;   // preallocated, ready to use.
+  interval_set<inodeno_t> delegated_inos;  // hand these out to client
```

Comment:
in-memory data structure shouldn't be defined in session_info_t.  delegated_inos should be defined in class Session.

Reply:
If we don't erase delegated_inos from prealloc_inos. just let delegated_inos be subset of prealloc_inos.  _session_logged related changes can be avoid.

Reply:
I think I tried that at one point, but that means that we have to end up doing allocations (take_ino) from the middle of an interval_set, and the interval_set code doesn't really support doing that well. I'll try again to be sure, but I think this is actually simpler (sadly).

Reply:
I saw warning removed by commit 'mds: don't warn when inodes are not allocated out sequentially' before.  For the warning case,  inode number is removed from middle of internal_set